clippy_lints/src/needless_continue.rs

```diff
@@ -361,24 +361,68 @@ fn suggestion_snippet_for_continue_inside_else(cx: &EarlyContext<'_>, data: &Lin
     )
 }
 
-fn check_and_warn(cx: &EarlyContext<'_>, expr: &ast::Expr) {
-    if_chain! {
-        if let ast::ExprKind::Loop(loop_block, ..) = &expr.kind;
-        if let Some(last_stmt) = loop_block.stmts.last();
-        if let ast::StmtKind::Expr(inner_expr) | ast::StmtKind::Semi(inner_expr) = &last_stmt.kind;
-        if let ast::ExprKind::Continue(_) = inner_expr.kind;
-        then {
-            span_lint_and_help(
-                cx,
-                NEEDLESS_CONTINUE,
-                last_stmt.span,
-                MSG_REDUNDANT_CONTINUE_EXPRESSION,
-                None,
-                DROP_CONTINUE_EXPRESSION_MSG,
-            );
+fn check_last_stmt<F>(b: &Block, func: &F)
+where
+    F: Fn(Option<&ast::Label>, Span),
+{
+    if let Some(last_stmt) = b.stmts.last() &&
+       let ast::StmtKind::Expr(inner_expr) | ast::StmtKind::Semi(inner_expr) = &last_stmt.kind {
+        match &inner_expr.kind {
+            rustc_ast::ExprKind::Continue(continue_label) => {
+                func(continue_label.as_ref(), last_stmt.span);
+            },
+
+            rustc_ast::ExprKind::If(_, then_block, else_block) => {
+                check_last_stmt(then_block, func);
+                if let Some(else_block) = else_block {
+                    match &else_block.kind {
+                        rustc_ast::ExprKind::Continue(continue_label) => {
+                            func(continue_label.as_ref(), else_block.span);
+                        }
+                        rustc_ast::ExprKind::Block(b, _) => {
+                            check_last_stmt(b, func);
+                        }
+                        _ => {}
+
+                    }
+                }
+            }
+            rustc_ast::ExprKind::Match(_, arms) => {
+                for arm in arms {
+                    match &arm.body.kind {
+                        rustc_ast::ExprKind::Continue(continue_label) => {
+                            func(continue_label.as_ref(), arm.body.span);
+                        }
+                        rustc_ast::ExprKind::Block(b, _) => {
+                            check_last_stmt(b, func);
+
+                        }
+                        _ => {}
+                    }
+
+                }
+            }
+            _ => {},
```

Comment:
Probably makes sense to check for `Block` here too? Would allow us to lint
```rs
loop {
  { continue; }
}
```
as well